### ISSUE TITLE
[DE300]: [nova] different schemes should be allowed for public, internal and admin endpoints

### ISSUE THREAD

**Issue by odyssey4me** [rcbops/chef-cookbooks#583]
Currently if you do the following for the environment then the public, internal and admin endpoints all become https:

``` json
      "services": {
        "api": {
          "scheme": "https"
        },
        "ec2-public": {
          "scheme": "https"
        }
```

Our ideal scenario is to make only the public endpoint SSL, with the others staying non-SSL. This ensures that the SSL overhead only comes into play when people are using the public API's. The thinking is that the internal services should rather use the internal endpoint without HTTPS, reducing the overhead.

**Comment by Apsu** [comment]
We actually have folks who want SSL internal endpoints as an option, as part of their layered security approach. I agree that we sould haven more granular options as to which endpoints get SSL and which do not, but we should no more assume that public is the only one we want than we should have assumed it was all or nothing :)

**Comment by odyssey4me** [comment]
Agreed, that's pretty much what I'm getting at. Each service should be able to have the scheme set individually.

The scenario I outlined, however, is our specific use-case which I'm trying to implement.

**Comment by andymcc** [comment]
I've been meaning to set this up for a while - it should be reasonably trivial since you can just "not" set up the VHost to use SSL, on a per endpoint basis.

**Comment by willkelly** [comment]
anymcc: I have put my thoughts below.  Based on your response, I think I'm missing something.

Using the current mod_wsgi approach, an individual service would have to bind to more than one ip/port combination in order to have the public endpoint be ssl and the others be plain text.  On most openstack services, this is not currently possible.

The other option is to always bind to plain text and to proxy requests through an ssl provider (apache, stunnel, stud, external loadbalancer).

**Comment by odyssey4me** [comment]
@willkelly If you take a look at how @andymcc did it in keystone you'll see that it should be possible. It's just a matter of using multiple VHosts.
